### PR TITLE
Feature/processor output individual records for gratia, add CPU usage and memory metrics

### PR DESCRIPTION
**Pull request by mwestphall**
- Add a CPU usage metric based on `container_cpu_usage_seconds_total`.
  - **NB**:  on old clusters using cgroups v1 this double counts results
- Add a new memory calculation using the `kube_pod_container_resource_requests{resource="memory"}` metric, similar to cores.
- Update KAPEL.py to optionally output individual APEL job records if gratia is being used for output. This aligns better with the expected GRACC record format
- Add a simple Dockerfile for building the processor InitContainer

@rptaylor The commit history is a bit messy for this change since I separated it out from the helm chart changes in-post, let me know if you'd like it condensed down with a rebase.

### CODE REVIEW

File: python/KAPEL.py [hunk 1]
```diff
@@ -52,7 +52,8 @@ def __init__(self, queryRange, namespace):
         # (which takes a range and returns a scalar), and as a result get the whole metric set. Finally, use group_left for many-to-one matching.
         # https://prometheus.io/docs/prometheus/latest/querying/operators/#aggregation-operators
         # https://prometheus.io/docs/prometheus/latest/querying/operators/#many-to-one-and-one-to-many-vector-matches
-        self.cputime = f'(max_over_time(kube_pod_completion_time{{namespace="{namespace}"}}[{queryRange}]) - max_over_time(kube_pod_start_time{{namespace="{namespace}"}}[{queryRange}])) * on (pod) group_left() max without (instance, node) (max_over_time(kube_pod_container_resource_requests{{resource="cpu", node != "", namespace="{namespace}"}}[{queryRange}]))'
+        self.cputime = f'max_over_time(container_cpu_usage_seconds_total{{namespace="{namespace}"}}[{queryRange}])'
```
Comment (rptaylor) [inline python/KAPEL.py:55]: @mwestphall  This looks like  CPU usage time which is very different from the current cputime (CPU walltime) query which should remain unchanged for APEL accounting to keep working. Having CPU usage would be useful though.

Based on some testing this query will double-count all records because it looks like you get container-level results along with the top-level cgroup (presumably?) corresponding to the pod. The only way I can see to only get the right top-level cgroup is to select records without the `name` field. See attached image as an example.

Comment (rptaylor) [reply python/KAPEL.py:55]: Screenshot from 2024-05-24 18-20-25

Comment (rptaylor) [reply python/KAPEL.py:55]: Hmm also these metrics are being collected from a different source, it must be the cAdvisor metrics built into the kubelet, instead of the kube-state-metrics pod. I wonder if you might run into different kinds of issues similar to those avoided by  using `without (instance, node)` as explained in the comments.

File: python/KAPEL.py [hunk 2]
```diff
@@ -52,7 +52,8 @@ def __init__(self, queryRange, namespace):
...
+        self.memory = f'max_over_time(container_memory_working_set_bytes{{namespace="{namespace}"}}[{queryRange}]) / 1000'
```
Comment (rptaylor) [inline python/KAPEL.py:56]: Likewise for this, it is double counting all memory usage.

File: python/KAPEL.py (hunk 1 above)
Comment (rptaylor) [reply python/KAPEL.py:55]: FYI this reports CPU usage for pods that are currently still running. A subsequent run would then report more usage for the same pod later.

File: README.md [hunk 3]
```diff
@@ -27,6 +27,8 @@ KAPEL is container-native Kubernetes accounting, for APEL and Gratia.
   - For large production deployments (examples are based on a cluster with about 125 nodes and 7000 cores):
     - Increase `.Values.prometheus.querySpec.timeout` (e.g. ~ 1800s) to allow long queries to succeed.
     - Apply sufficient CPU and memory resource requests and limits.
+  - If Prometheus is configured to require authentication, configure `.Values.processor.prometheus_auth` to specify an appropriate kubernetes
```
[the remark this answers is not in the document]
Comment (mwestphall) [reply README.md]: Oops, looks like this change snuck back in while I was resolving merge conflicts. I reverted it back out.

File: python/KAPEL.py (hunk 1 above)
Comment (mwestphall) [reply python/KAPEL.py:55]: @rptaylor Good catch with the double counting of records, I didn't catch that. It looks like kubernetes creates a separate ["pause" container](https://kubernetes.io/docs/concepts/windows/intro/#pause-container) within each pod to hold the cgroup in case the other containers crash, that's likely what's leading to the duplicate records here. With regards to deduplication, do you think it would work to do a `by (pod)` rather than `without (<list of attributes>)`? In our prometheus instance, the record for the pause container differs from the record for the main container by several attributes (`without(name, id, image, namespace, container)` appears to deduplicate successfully)
image

Comment (rptaylor) [reply python/KAPEL.py:55]: I think `by pod` will hide the double counting without fixing it: https://stackoverflow.com/questions/51064821/difference-between-promql-by-and-without-unclear
Using by or without are different ways of aggregating results together, not deduplicating per se.
AFAICT based on the results I saw, the only way to be sure to count only the record corresponding to the parent cgroup (i.e. the overall pod itself without also including the component containers) would be to select items without a name field (maybe `name == ""` or `name=Null` would work? Not sure of the syntax), or parse the id field to make sure it matches the form of the parent cgroup of a pod (though yours looks different from mine, maybe it is a cgroups v1 vs v2 difference).

Another option could be to select only records that have a defined image field (i.e. all cgroups *except* the parent pod). Using without or by would make sense here to aggregate all those records, one for each real container, into one result..

Comment (mwestphall) [reply python/KAPEL.py:55]: Just for clarification, the record that we want to report on is (in the example screenshot you sent) the second record with no name attribute? In that scenario, would attempting to use `max by (pod)` rather than `sum by (pod)` work, since it'd only take the highest CPU time from the nameless record rather than the lesser values from the other two? 

To further complicate things, which versions of kubernetes is your cluster using? In Tiger, prometheus is not reporting that nameless record, just one for the pause container and one for the pod's main container. Tiger is on version kubernetes 1.27, where it looks like `container_cpu_usage_seconds_total` is still an alpha feature. It doesn't become stable until 1.29. 
image

Comment (rptaylor) [reply python/KAPEL.py:55]: Currently our clusters are v1.24 and as I understand cgroup v2 is promoted to stable in v1.25, so yes that could explain the difference.

Yes I think what you want is the CPU usage of the whole pod (all containers), whereas by its name `container_cpu_usage_seconds_total` is a container-based metric.
My interpretation of the screenshot I posted is: 13.14 for the main job container in the pod, .03 for the pause container, 13.16 total for the whole pod.   Some pods could have multiple workload containers.

My understanding of `max` is it selects the larger of two different metrics but in this case the only metric you're interested in is container_cpu_usage_seconds_total. I think the issue here is one metric, with multiple time series data points, distinguished by different labels. 

Hmm so maybe in cgroups v2, if the pod itself is not (mis)reported along with the containers in the container_cpu_usage_seconds_total metric, there should not be an issue of double counting. If so, based on your latest screenshot, if for every pod you only get the child containers, then adding them all together would be correct.

Comment (mwestphall) [reply python/KAPEL.py:55]: Still looking into this, but the query that I ran a couple hours ago no longer returns any results. I believe there may be an additional issue that the metrics derived from cAdvisor aren't persisted in tiger for as long as those from kube-state-metrics, which will definitely cause additional issues.

Comment (mwestphall) [reply python/KAPEL.py:55]: That query is now returning results again. I will definitely need to investigate the cause of that data absence though. It looks like you can aggregate `by` a single metric to get the largest of any number of records, grouped by the `by` parameter:
```
max_over_time(container_cpu_usage_seconds_total{namespace="uwdf-dev", pod="kapel-test-ep-rs62w"}[1d])

{container="osgvo-docker-pilot", cpu="total", endpoint="https-metrics", image="hub.opensciencegrid.org/opensciencegrid/osgvo-docker-pilot:23-release", namespace="uwdf-dev", node="tiger0011.chtc.wisc.edu", pod="kapel-test-ep-rs62w", service="kubelet", ...}
	49.119534462
{cpu="total", endpoint="https-metrics", image="k8s.gcr.io/pause:3.2", namespace="uwdf-dev", node="tiger0011.chtc.wisc.edu", pod="kapel-test-ep-rs62w", service="kubelet", ...}
	0.037396571
```

vs
```
max by (pod) max_over_time(container_cpu_usage_seconds_total{namespace="uwdf-dev", pod="kapel-test-ep-rs62w"}[1d])

{pod="kapel-test-ep-rs62w"}
	49.119534462
```
The `max by` takes the largest of any `container_cpu_usage_seconds_total` records where `pod = "kapel-test-ep-rs62w"`. This should give an accurate value in your cluster version since the whole-pod metric is always going to be the largest, but will slightly undercount in version 1.27 since it's excluding the pause container's small amount of cpu usage.

Comment (rptaylor) [reply python/KAPEL.py:55]: I also noticed some irregularities regarding the time periods for which container_cpu_usage_seconds_total metrics for a given job were recorded, but couldn't narrow it down. Also if Prometheus is ever unavailable for a period of time you get a gap in the records.

In principle, some pods could have multiple worker containers and it would be good to account all of them, unless it is documented as a limitation of the Gratia CPU usage reporting in Kuantifier.

In any case the original point of this thread remains, this is a very different query than the CPU time needed by APEL. Can you make a new separate variable for this query?  Having both CPU wall time and CPU usage will be useful.

Comment (mwestphall) [reply python/KAPEL.py]: @rptaylor oops, I made that change last week but never pushed it. The original metric should now be restored with a separate metric for the container_cpu_usage_seconds_total metric.

Comment (rptaylor) [reply python/KAPEL.py]: From what I understood previously https://github.com/rptaylor/kapel/pull/54#discussion_r1619323129, I thought with cgroups v2, it turned out there would not be an issue of double counting after all? (The reason I thought there would is because of the cgroups v1 results I was seeing). You get two results, one for the pause container and one for the work container, and if you aggregate them together you get the right amount. I think it should be like this:
`sum by (pod) (max_over_time(container_cpu_usage_seconds_total{namespace="harvester",pod="grid-job-5143931-glsmj"}[1d]))`

If you do `max by (pod)` it means that only the largest CPU user of all containers in a pod will be accounted, so there would need to be a documented limitation that the results are not correct if a pod has multiple worker containers. This would make the accounting system unusable for any new job workflows that can take advantage of e.g. sidecar patterns in k8s, which are not possible in traditional batch systems.

In other words, if there is going to be a known limitation, I would rather it be that you get double counting if and only if the old cgroups v1 is used (i.e. using sum by pod), rather than only counting one container in a pod which uses the most CPU (i.e. using max by pod) in the modern cgroups v2. Does that make sense?

Comment (mwestphall) [reply python/KAPEL.py]: @rptaylor yeah, that sounds good. `sum by (pod)` should be more accurate for v2 cgroups, `max by (pod)` was just an attempt to compromise between the v1 and v2 reporting. I'll update to use `sum by (pod)`.

Comment (rptaylor) [reply python/KAPEL.py]: Nice. I noticed one other improvement, since the CPU usage should be a monotonically increasing cumulative value, you should be able to use `last_over_time` instead of max_over_time which should be much easier performance-wise on your Promethei, especially for long jobs. Based on my testing I confirmed the results look the same and for large time periods it was much faster.

https://prometheus.io/docs/prometheus/latest/querying/functions/#aggregation_over_time

File: python/KAPEL.py [hunk 4]
```diff
@@ -53,6 +53,12 @@ def __init__(self, queryRange, namespace):
         # https://prometheus.io/docs/prometheus/latest/querying/operators/#aggregation-operators
         # https://prometheus.io/docs/prometheus/latest/querying/operators/#many-to-one-and-one-to-many-vector-matches
         self.cputime = f'(max_over_time(kube_pod_completion_time{{namespace="{namespace}"}}[{queryRange}]) - max_over_time(kube_pod_start_time{{namespace="{namespace}"}}[{queryRange}])) * on (pod) group_left() max without (instance, node) (max_over_time(kube_pod_container_resource_requests{{resource="cpu", node != "", namespace="{namespace}"}}[{queryRange}]))'
+
+        # These are container-level memory and CPU usage metrics reported by kubelets.
+        self.memory = f'sum by (pod) (max_over_time(container_memory_working_set_bytes{{namespace="{namespace}"}}[{queryRange}])) / 1000'
```
Comment (rptaylor) [inline python/KAPEL.py:58]: Unlike the container_cpu_usage_seconds_total which is a cumulative counter, the memory usage is an instantaneous gauge. For a pod with containers that have memory usage like this:
```
                                  minute 1              minute 2
container 1                    1 GB                          0 GB
container 2                    0 GB                          1 GB
```
the query would incorrectly report memory usage of 2 GB even though the real memory usage never exceeded 1 GB, because it is doing the sum of max_over_time of each. I tried to find a way to collapse together the range vector for all containers in a pod first, then take the max but couldn't figure it out. 

@mwestphall What would you think of using the combined resource memory request of all containers in a pod, like `sum by (pod) (max_over_time(kube_pod_container_resource_requests{resource="memory", node != "", namespace="harvester"}[1h]))` ?  This uses metrics from kube-state-metrics, which should be simpler and more reliable than the cgroup-based ones.

Besides, the requested memory is the usage in the sense of what a pod deprives other pods of using. The actual memory occupied by a process could be lower than what is requested, but the pod still "used" all the requested memory in the sense that it was reserved by the scheduler.

Comment (mwestphall) [reply python/KAPEL.py:58]: @rptaylor Using the `kube_pod_container_resource_requests` metric in place of `container_cpu_usage_seconds_total` sounds good to me. 

Then with regards to multi-container support in general, we ran into an issue when we were first testing on tiger with the cputime query erroring out when run against multi-container pods. I believe this is because the `max without (instance, node)` grouping in the query doesn't always aggregate every container in a pod down to a single record. Would it make sense to a `max by (pod)` there instead?

Comment (rptaylor) [reply python/KAPEL.py:58]: Do you mean using e.g. `sum by (pod) (max_over_time(kube_pod_container_resource_requests{resource="memory", node != "", namespace="harvester"}[1h]))` instead of container_memory_working_set_bytes ?

I think the container_cpu_usage_seconds_total query should be good now, after all the investigation we did here.

I opened a separate issue for that: https://github.com/rptaylor/kapel/issues/56

Comment (rptaylor) [reply python/KAPEL.py:58]: I thought your commit "[report on pod total requested memory rather than max usage per container](https://github.com/rptaylor/kapel/pull/54/commits/a0e56e706bce6906b2958534f3d39431345e91b3)" was meant to address this but it looks like it only changes the chart name.

Comment (mwestphall) [reply python/KAPEL.py]: Oops, good catch, that was definitely not the intended change with that commit. I've rebased that out and changed the `memory` query to `sum by (pod) (max_over_time(kube_pod_container_resource_requests{resource="memory", node != "", namespace="harvester"}[<queryRange>]))` as intended.

File: python/KAPEL.py [hunk 5]
```diff
@@ -262,6 +262,73 @@ def process_period(config, period):
...
+    print(
+        f"Processing year {period['year']}, month {period['month']}, "
+        f"querying from {period['instant'].isoformat()} and going back {period['range_sec']} s to {period_start.isoformat()}."
+    )
+    queries = QueryLogic(queryRange=(str(period['range_sec']) + 's'), namespace=config.namespace)
+
+    # SSL generally not used for Prometheus access within a cluster
+    # Docs on instant query API: https://prometheus.io/docs/prometheus/latest/querying/api/#instant-queries
+    headers = {"Authorization": config.auth_header } if config.auth_header else None
+    prom = PrometheusConnect(url=config.prometheus_server, disable_ssl=True, headers=headers)
```
Comment (rptaylor) [inline python/KAPEL.py:310]: Previously the assumption was using unauthenticated HTTP inside the cluster.
But if you're passing an auth header you probably want to use HTTPS to keep it secure. 
However whether or not the URL actually specifies HTTPS is up to the configurable value set for PROMETHEUS_SERVER.
So maybe the best thing would be to [check the scheme of the URL](https://github.com/sloria/environs?tab=readme-ov-file#url-schemes) and set `disable_ssl` only if the scheme is HTTP rather than HTTPS. What do you think @mwestphall ?

File: python/KAPEL.py [hunk 6]
```diff
@@ -81,6 +87,29 @@ def summary_message(config, year, month, wall_time, cpu_time, n_jobs, first_end,
...
+    output = (
+        f'APEL-individual-job-message: v0.3\n'
+        f'Site: {config.site_name}\n'
+        f'VO: {config.vo_name}\n'
+        f'SubmitHost: {config.submit_host}\n'
+        f'MachineName: {pod_name}\n'
+        f'LocalJobId: {pod_name}\n'
+        f'InfrastructureType: {config.infrastructure_type}\n'
+        f'ServiceLevelType: si2k\n'
+        f'ServiceLevel: {config.benchmark_value * 250}\n'
```
Comment (rptaylor) [inline python/KAPEL.py:101]: This would oblige Kuantifier users in GRACC mode to put in some dummy benchmark value, although GRACC won't read it. I think it would be better to omit than to require a value that won't be used. I would be okay with simply commenting out ServiceLevelType and ServiceLevel here since the individual messages are not used in APEL mode, or maybe something like
```
benchmark_output = those two lines of f strings if config.benchmark_value is defined, else empty string
```
and then include `benchmark_output` in the printing of `output`. What do you think @mwestphall  ?

File: python/KAPEL.py [hunk 7]
```diff
@@ -159,43 +188,14 @@ def rearrange(x):
...
-        del raw_results[query_name]
 
+def record_summarized_period(config, period_start, year, month, results):
+    """ Record the sum of usage across all pods in the given time period. """
     cputime = results['cputime']
     endtime = results['endtime']
     starttime = results['starttime']
     cores = results['cores']
-
+    result_lengths = max(len(l) for l in results.values())
```
Comment (rptaylor) [inline python/KAPEL.py:198]: @mwestphall Previously result_lengths was an array containing the number of results of the cputime, starttime, endtime, cores queries (and now also cpuusage and memory).
So shouldn't result_lengths be perhaps something like this?
`result_lengths = [(len(l) for l in results.values())]`   (I didn't test the syntax.)

Otherwise the assertion on line 203 might fail because first it would take the max of the array but then it is checked against the min of the array.

Comment (mwestphall) [reply python/KAPEL.py:198]: @rptaylor good catch here, the `min(max([<array>]))` that's effectively being done here gives a TypeError. The change you suggested should work.

File: python/KAPEL.py (hunk 6 above)
Comment (mwestphall) [reply python/KAPEL.py:101]: I would agree, but think that splitting out the GRACC and APEL specific config values might make sense to address in a separate PR. `KAPELConfig.py` will also currently throw an error if `BENCHMARK_VALUE` is non-null.

File: python/KAPEL.py (hunk 5 above)
Comment (mwestphall) [reply python/KAPEL.py:310]: We might want to make disable_ssl configurable directly at the values.yaml level, I think its usage will vary case-by-case. It looks like it ultimately [sets `requests.Session().validate`](https://github.com/4n4nd/prometheus-api-client-python/blob/c77c0b91d85a709848e72b457ef26b706a1ceccb/prometheus_api_client/prometheus_connect.py#L77), which [only applies for https requests](https://requests.readthedocs.io/en/latest/user/advanced/#ssl-cert-verification) and not http. In tiger, we do all our ssl termination at the ingress level so the Prometheus service would not accept https connections at the internal service address. This will probably vary cluster to cluster though.

Comment (rptaylor) [reply python/KAPEL.py:310]: Good point, normally TLS happens at cluster ingress. I thought if you're authenticating inside the cluster there must be TLS inside the cluster. Maybe sending the credential in plain text inside the cluster isn't a big concern. Anyway there could be a separate issue for that if needed.

File: python/KAPEL.py (hunk 6 above)
Comment (rptaylor) [reply python/KAPEL.py:101]: Okay sure, a separate PR sounds good.

**Review by rptaylor — APPROVED** [review]
LGTM

**Comment by rptaylor** [comment]
Actually for the record I believe the cgroup-related differences we saw are due to using EL8 where the default is still cgroup v1, not because of the k8s version per se.
